### PR TITLE
fix: first data of the epoch from dataset is inappropriate

### PR DESCRIPTION
**Pull request by kohya-ss**
This pull request fixes the issue where the first data of each epoch is retrieved by the wrong bucket index.

Each dataset shuffles ARB (aspect ratio bucketing) buckets at the epoch change. In the current implementation, the epoch change is notified by collate_fn, that is, after the first data is retrieved from the dataset. Therefore, the first data is retrieved based on the bucket order before shuffling. This will result in a maximum of one duplicate or missing (not trained) of data within one epoch.

In the new implementation, the epoch number in the shared memory is updated before starting to read from DataLoader for the epoch. Also, each dataset checks the epoch number before acquiring data at each step. This means that when acquiring the first data of a certain epoch, each dataset knows that the epoch has been updated.

### CODE REVIEW

File: src/musubi_tuner/dataset/image_video_dataset.py [hunk 1]
```diff
@@ -1431,7 +1426,16 @@ def __len__(self):
         return NotImplementedError
 
     def __getitem__(self, idx):
-        raise NotImplementedError
+        epoch = self.shared_epoch.value
```
[the remark this answers is not in the document]
Comment (kohya-ss) [reply src/musubi_tuner/dataset/image_video_dataset.py:1431]: I've updated.

**Comment by kohya-ss** [comment]
The PR fixes the edge case where the very first sample of an epoch could follow the previous epoch’s order. When viewed across multiple epochs, each sample is ultimately consumed exactly once per epoch cycle—no missing or duplicated usage—so data coverage and gradient statistics are effectively unchanged. The primary effect is improved **in-epoch** consistency at boundaries; under identical settings, longer-horizon training behavior (metrics/convergence) should remain essentially the same. *(In extremely short runs or with early stopping, minor differences from ordering effects may still be observable.)*

 **Summary**

* We now announce epoch transitions **before** entering the DataLoader fetch loop via `shared_epoch (Synchronized[int])`.
* **Right after detecting a change in `shared_epoch` inside `__getitem__`**, the dataset shuffles ARB buckets. This ensures the **first sample of the epoch already follows the new order**, eliminating duplicate/missed samples within an epoch.
* **Reproducibility note:** Because the shuffle timing is earlier, **the exact sample order will not match the old implementation even with the same seed** (reproducibility is preserved within the same implementation).

**Detail**

## Background / Issue

Previously, the epoch-change notification happened **after the very first sample of an epoch was fetched** (via the collate path). As a result, the **first sample of each epoch could be taken using the previous epoch’s ARB bucket order**, which could lead to **at most one duplicate or one missed sample per epoch**.

## What This PR Changes

* **Shared epoch indicator:** Introduces `shared_epoch: Synchronized[int]`. The main training script writes the new epoch number **at the beginning of the epoch, right before entering the DataLoader’s data-fetch loop**.
* **Early shuffle:** Each dataset **checks `shared_epoch` inside `__getitem__`**; **if the value changed, it immediately shuffles ARB buckets** before returning a sample.
* This guarantees that **from the very first sample of the new epoch**, the correct (new) ARB order is applied, eliminating the duplicate/missed-sample edge case.

## Trigger & Update Points

* **Trigger:** **Shuffle is executed immediately after `__getitem__` detects a change in `shared_epoch`**.
* **Update point:** `shared_epoch` is updated **in the main training script at the start of each epoch**, right before the DataLoader begins fetching samples.

## Reproducibility Notes

* Because the shuffle now occurs earlier, **the exact sample consumption order no longer matches the old implementation even with the same seed**.
* **Within the new implementation**, runs remain reproducible given the **same seed and identical settings**.
* If you rely on historical logs or comparisons that assume the old ordering, expect differences.

## Impact

* Sample consumption becomes **consistent across the entire epoch** (from the first sample), removing the **“up to one duplicate/missed sample per epoch”** behavior.
